src/Storages/ObjectStorageQueue/ObjectStorageQueueMetadata.cpp

```diff
@@ -687,63 +686,145 @@ size_t ObjectStorageQueueMetadata::unregisterActive(const StorageID & storage_id
     return remaining_nodes_num;
 }
 
-size_t ObjectStorageQueueMetadata::unregisterNonActive(const StorageID & storage_id)
+size_t ObjectStorageQueueMetadata::unregisterNonActive(const StorageID & storage_id, bool remove_metadata_if_no_registered)
 {
     const auto registry_path = zookeeper_path / "registry";
+    const auto drop_lock_path = zookeeper_path / "drop";
     const auto self = Info::create(storage_id);
 
     Coordination::Error code = Coordination::Error::ZOK;
+
     for (size_t i = 0; i < 1000; ++i)
     {
-        Coordination::Stat stat;
-        std::string registry_str;
-        auto zk_client = getZooKeeper();
+        Coordination::Requests requests;
+        Coordination::Responses responses;
+        size_t count = 0;
 
-        bool node_exists = zk_client->tryGet(registry_path, registry_str, &stat);
-        if (!node_exists)
+        bool supports_remove_recursive = true;
+        zkutil::ZooKeeperPtr zk_client;
+
+        try
         {
-            LOG_WARNING(log, "Cannot unregister: registry does not exist");
-            chassert(false);
-            return 0;
-        }
+            zk_client = getZooKeeper();
```

Comment:
Here part of old code was just moved inside try-catch, because `trySet/tryGet` actually throw an exception in case of "keeper session expired error" instead of returning the error code like for other error types, therefore this error was not retried before.